addOns/client/src/main/java/org/zaproxy/addon/client/internal/ClientMapWriter.java

```diff
@@ -0,0 +1,202 @@
+                            .getExtension(ExtensionClientIntegration.class);
+            exportClientMap(fw);
+        }
+    }
+
+    public static void exportClientMap(Writer fw) throws IOException {
+        exportClientMap(fw, extension.getClientTree());
+    }
+
+    public static void exportClientMap(Writer fw, ClientMap clientMap) throws IOException {
+        try (BufferedWriter bw = new BufferedWriter(fw)) {
+            outputNode(bw, clientMap.getRoot(), 0);
+        }
+    }
+
+    private static boolean outputKV(
+            Writer fw, String indent, boolean first, String key, Object value) throws IOException {
+        fw.write(indent);
+        if (first) {
+            fw.write("- ");
+        } else {
+            fw.write("  ");
+        }
+        fw.write(key);
+        fw.write(": ");
+        ObjectMapper mapper =
+                new ObjectMapper(
+                        new YAMLFactory()
+                                .enable(Feature.LITERAL_BLOCK_STYLE)
+                                .disable(Feature.WRITE_DOC_START_MARKER));
+        // For some reason the disable start marker doesn't seem to work
+        String output = mapper.writeValueAsString(value).replace("--- ", "");
+        fw.write(output);
+        return false;
+    }
+
+    private static void outputNode(Writer fw, ClientNode node, int level) throws IOException {
+        if (node.isStorage()) {
+            // Skip storage nodes in the tree
+            // Those details are represented as components of their source
+            return;
+        }
+        // We could create a set of data structures and use jackson, but the format is very
+        // simple and this is much more memory efficient - it still uses jackson for value
+        // output
+        String indent = " ".repeat(level * 2);
+
+        outputKV(
+                fw,
+                indent,
+                true,
+                NODE_KEY,
+                level == 0 ? ROOT_NODE_NAME : node.getUserObject().getName());
+
+        if (node.getUserObject().isStorage()) {
+            outputKV(fw, indent, false, STORAGE_KEY, node.getUserObject().isStorage());
+        }
+        if (!node.getUserObject().isVisited()) {
+            outputKV(fw, indent, false, VISITED_KEY, node.getUserObject().isVisited());
+        }
+        if (node.getUserObject().getComponents() != null
+                && !node.getUserObject().getComponents().isEmpty()) {
+            for (ClientSideComponent component : node.getUserObject().getComponents()) {
+                if (component.getTypeForDisplay().equalsIgnoreCase(ClientSideComponent.REDIRECT)) {
+                    outputKV(
+                            fw,
+                            indent,
+                            false,
+                            ClientSideComponent.REDIRECT.toLowerCase(Locale.ROOT),
+                            component.getHref());
+                }
+                if (component
```

Comment:
`else if`, it can be both at the same time.

Reply:
⭐